src/Datastore/DatastoreClient.php

```diff
@@ -0,0 +1,907 @@
+            'ancestors' => [],
+            'id' => null,
+            'name' => null
+        ];
+
+        $path = [];
+        if (count($options['ancestors']) > 0) {
+            $path = $options['ancestors'];
+        }
+
+        $path[] = array_filter([
+            'kind' => $kind,
+            'id' => $options['id'],
+            'name' => $options['name']
+        ]);
+
+        $key = new Key($this->projectId, [
+            'path' => $path,
+            'namespaceId' => $options['namespaceId']
+        ]);
+
+        $keys = array_fill(0, $options['number'], $key);
+
+        if ($options['allocateIds']) {
+            $keys = $this->allocateIds($keys);
+        }
+
+        return $keys;
+    }
+
+    /**
+     * Create an entity.
+     *
+     * This method does not execute any service requests.
+     *
+     * ```
+     * $key = $datastore->key('Person', 'Bob');
+     * $entity = $datastore->entity($key, [
+     *     'firstName' => 'Bob',
+     *     'lastName' => 'Testguy'
+     * ]);
+     * ```
+     *
+     * ```
+     * // Entities can be manipulated via PHP's arrayaccess:
+     * $entity['firstName'] = 'Bob';
+     * $entity['lastName'] = 'Testguy';
+     *
+     * echo $entity['firstName']; // 'Bob'
+     * ```
+     *
+     * @see https://cloud.google.com/datastore/reference/rest/v1beta3/Entity Entity
+     *
+     * @param Key $key The Key used to uniquely identify this entity. Only
+     *        complete keys may be used.
+     * @param array $entity The data to fill the entity with.
+     * @param array $options See {@see Google\Cloud\Datastore\Entity::__construct()} for documentation.
+     * @return Entity
+     */
+    public function entity(Key $key, array $entity = [], array $options = [])
+    {
+        return new Entity($key, $entity, $options);
+    }
+
+    /**
+     * Allocates an available ID to a given incomplete key
+     *
+     * Key MUST be in an incomplete state (i.e. including a kind but not an ID
+     * or name in its final pathElement).
+     *
+     * This method will execute a service request.
+     *
+     * Example:
+     * ```
+     * $key = $datastore->key('Person');
+     * $allocatedId = $datastore->allocateId($key);
+     * ```
+     *
+     * @see https://cloud.google.com/datastore/reference/rest/v1beta3/projects/allocateIds allocateIds
+     *
+     * @param Key $key The incomplete key
+     * @param array $options Configuration options
+     * @return Key
+     */
+    public function allocateId(Key $key, array $options = [])
+    {
+        $res = $this->allocateIds([$key], $options);
+        return $res[0];
+    }
+
+    /**
+     * Allocate available IDs to a set of keys
+     *
+     * Keys MUST be in an incomplete state (i.e. including a kind but not an ID
+     * or name in their final pathElement).
+     *
+     * This method will execute a service request.
+     *
+     * Example:
+     * ```
+     * $keys = $datastore->keys('Person', ['number' => 2]);
```

Comment:
I've been back and forth on whether `allocateIds` should be enabled by default. WDYT? My guess is that it would usually be used, but I'm a little worried about opting in to a service request that might surprise people a bit.